kombu/transport/redis.py

```diff
@@ -179,6 +181,91 @@ def _after_fork_cleanup_channel(channel):
     channel._after_fork()
 
 
+class PrefixedStrictRedis(redis.StrictRedis):
+    """Returns a ``StrictRedis`` client that prefixes the keys it uses.
+
+    Overriding all the methods used by Kombu with the same key prefixing logic
+    would be cumbersome and inefficient. Hence, we override the command
+    execution logic that is called by all commands.
+    """
+
+    PREFIXED_SIMPLE_COMMANDS = [
+        "HDEL",
+        "HGET",
+        "HSET",
+        "LLEN",
+        "LPUSH",
+        "PUBLISH",
+        "SADD",
+        "SET",
+        "SMEMBERS",
+        "ZADD",
+        "ZREM",
+        "ZREVRANGEBYSCORE",
+    ]
+
+    PREFIXED_COMPLEX_COMMANDS = {
+        "BRPOP": {"args_start": 0, "args_end": -1},
+        "EVALSHA": {"args_start": 2, "args_end": 3},
+    }
+
+    def __init__(self, *args, **kwargs):
+        self.global_keyprefix = kwargs.pop('global_keyprefix', '')
+        super().__init__(self, *args, **kwargs)
+
+    def _prefix_args(self, args):
+        args = list(args)
+        command = args.pop(0)
+
+        if command in self.PREFIXED_SIMPLE_COMMANDS:
+            args[0] = self.global_keyprefix + str(args[0])
+
+        if command in self.PREFIXED_COMPLEX_COMMANDS.keys():
+            args_start = self.PREFIXED_COMPLEX_COMMANDS[command]["args_start"]
+            args_end = self.PREFIXED_COMPLEX_COMMANDS[command]["args_end"]
+
+            pre_args = args[:args_start] if args_start > 0 else []
+
+            if args_end is not None:
+                post_args = args[args_end:]
+            elif args_end < 0:
+                post_args = args[len(args):]
+            else:
+                post_args = []
+
+            args = pre_args + [
+                self.global_keyprefix + str(arg)
+                for arg in args[args_start:args_end]
+            ] + post_args
+
+        return [command, *args]
+
+    def execute_command(self, *args, **kwargs):
+        return super().execute_command(*self._prefix_args(args), **kwargs)
+
+    def pipeline(self, transaction=True, shard_hint=None):
+        return PrefixedRedisPipeline(
+            self.connection_pool,
+            self.response_callbacks,
+            transaction,
+            shard_hint,
+            global_keyprefix=self.global_keyprefix,
+        )
+
+
+class PrefixedRedisPipeline(PrefixedStrictRedis, redis.client.Pipeline):
+    """Custom Redis pipeline that takes global_keyprefix into consideration.
+
+    As the ``PrefixedStrictRedis`` client uses the `global_keyprefix` to prefix
+    the keys it uses, the pipeline called by the client must be able to prefix
+    the keys as well.
+    """
+
+    def __init__(self, *args, **kwargs):
+        self.global_keyprefix = kwargs.pop('global_keyprefix', '')
+        redis.client.Pipeline.__init__(self, *args, **kwargs)
```

Comment:
can we use super() call here instead?

Reply:
I don't think we can as this class is using multiple inheritance.
I'm not sure why it has to be both a client and a pipeline.

Reply:
@thedrow is right, we cannot use only `super()` call here as far as I remember correctly.

The thing is that [`Pipeline`](https://github.com/andymccurdy/redis-py/blob/master/redis/client.py#L4245) inherits from `Redis`. Both has the `execute_command` method which must be overridden to prefix the keys it is using. We could do three things here as far as I understand:

1. Create a mixin that overrides the `execute_command` method and use that for both the pipeline and the custom redis class
2. Reimplement the exact same key prefixing logic in both pipeline and custom redis classes
3. Do what I did and create the custom pipeline using the original Pipeline class and the already patched redis class (`PrefixedStrictRedis`) and override the init function to call only the Pipeline's init

Probably we have other options here too, so if you would prefer something else, please let me know and I'll adjust the PR.

cc: @auvipy

Reply:
I think that a mixin would be far more obvious.
This code confused me as I was unaware of the fact that a Pipeline inherits from StrictRedis.

Reply:
@thedrow Right, I'll refactor to have a mixin instead.

Reply:
@thedrow the mixin was introduced as part of df8dfdc91b4ea5ed0a1b50c9f6cee01216d6e607